server/src/main/java/org/opensearch/cluster/routing/IndexShardRoutingTable.java

```diff
@@ -317,6 +329,29 @@ public ShardIterator activeInitializingShardsWeightedIt(WeightedRouting weighted
             List<ShardRouting> orderedInitializingShards = getInitializingShardsByWeight(weightedRouting, nodes, defaultWeight);
             ordered.addAll(orderedInitializingShards);
         }
+
+        // append shards for attribute value with weight zero, so that shard search requests can be tried on
+        // shard copies in case of request failure from other attribute values.
+        if (isFailOpenEnabled) {
+            try {
+                Stream<String> keys = weightedRouting.weights()
+                    .entrySet()
+                    .stream()
+                    .filter(entry -> entry.getValue().intValue() == 0)
+                    .map(Map.Entry::getKey);
+                keys.forEach(key -> {
+                    ShardIterator iterator = onlyNodeSelectorActiveInitializingShardsIt(weightedRouting.attributeName() + ":" + key, nodes);
+                    while (iterator.remaining() > 0) {
+                        ordered.add(iterator.nextOrNull());
+                    }
+                });
+            } catch (IllegalArgumentException e) {
+                // this exception is thrown by {@link onlyNodeSelectorActiveInitializingShardsIt} in case count of shard
+                // copies found is zero
+                logger.debug("no shard copies found for shard id [{}] for node attribute with weight zero", shardId);
+            }
+        }
+
```

Comment:
I just realised we aren't maintaining any state across request and hence we might just be implementing an incorrect version of weighted routing.

Reply:
State is maintained through across requests using seed https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/cluster/routing/IndexShardRoutingTable.java#L312

But since the shuffler sits in the IndexShardRoutingTable class, the state would be lost in case of modifications to `IndexShardRoutingTable`

Created a follow up task for this https://github.com/opensearch-project/OpenSearch/issues/5754